vo: add `video-target-params` property

kasper93 opened the pull request:
EDIT: removed images, because it is evolving.

kasper93 commented:
I'm not happy with the amount of mpv<->libplacebo glue it requires. And in general the duplication of code between both is really noticeable. Probably it is not a topic for this PR, but we could consider making libplacebo core dependency and at least use more robust colorimetric structs and avoid duplication/mapping of them. //CC @haasn

kasper93 commented:
> Although not relevant to this PR, I think there are too many spaces next to the HDR metadata.

It is standard indentation after prefix in stats.lua, I don't see better way to handle it, single space would be too little. This works better with all the fonts and so on, I think.

kasper93 commented:
> `color-depth` seems to only work with video-target-params, `video-params/color-depth` is always unavailable

Expected, for `video-params` use `average-bpp`. Supporting `mp_bit_encoding` fully is out of scope of this PR.

kasper93 commented:
> If is a non-PQ video and `target-trc` is not set, `video-target-params/gamma` is `auto`, is this the expected behavior?

I've seen it too. Probably should be set, but it comes from libplacebo. I think for consistency libplacebo should set all target parameters to which it inferred internally.

kasper93 commented:
@arm64-v9a: You can test with https://code.videolan.org/videolan/libplacebo/-/merge_requests/557 and this diff

``` diff
diff --git a/video/out/vo_gpu_next.c b/video/out/vo_gpu_next.c
index a55612f088..265fcf2547 100644
--- a/video/out/vo_gpu_next.c
+++ b/video/out/vo_gpu_next.c
@@ -1120,7 +1120,11 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
     }
 
     // Render frame
+#if PL_API_VER >= 315
+    if (!pl_render_image_mix_ex(p->rr, &mix, &target, &params)) {
+#else
     if (!pl_render_image_mix(p->rr, &mix, &target, &params)) {
+#endif
         MP_ERR(vo, "Failed rendering frame!\n");
         goto done;
     }
```

kasper93 commented:
> Although not relevant to this PR, I think there are too many spaces next to the HDR metadata.

After looking at it indeed it was bit much. Reduced a little, but still with a gap, so it is not too crowded. I don't know ASS well, probably this could be aligned in some nicer table like layout, but dunno.

llyyr commented:
> Unrelated but you did say "ASS" and I was thinking you meant to add subtitle info inside stats.lua too, so this could be an idea...

stats.lua, the osd as well as the osc are in ASS and rendered with libass

kasper93 commented:
> so why not subtitles stats too?

There is not much interesting things to be shown about current selected subtitle. Well maybe, language and type. It is possible, but I think noone really though it would be useful before.

Hrxn commented:
But now *Video Out* is between *Video* and *Audio*.. and they both belong together imho, as properties of a file..

> How about adding Picture Type and Frame number? Very useful for anime screenshots

Why only anime screenshots, though? 🤔

kasper93 commented:
> I feel it would be more aesthetically pleasing to put Video Out under Video?

I see your point, but as already pointer Video and Audio should be together. Else we have weird mess. If something like `Audio Out` is added, we can reorder it for symmetry, but now I would like to keep current order.

kasper93 commented:
> How about adding Picture Type and Frame number? Very useful for anime screenshots

It may sound strange from me, because I added to this page a lot of things that some may consider useless, but I also want to keep it still readable or at least not too busy. Maybe it is too late, but we can't add thing endlessly.

But I agree frame number would actually be useful. I will try to make it in one line, something like `Frame: I (12/123)` or something. Need to see what looks nice.

llyyr commented:
Here's my bikesheeding contribution since we're adding many numbers to stats.lua and modifying the layout a bit:

- Consider renaming "Video Out" to just Display or something
- Maybe only list specified and estimated display fps and file fps if they differ?
- I don't really think it's necessary to show bit depth information for pixel formats. You can already deduce this just from the pixel format itself, so if the user can't then "bpp" isn't going to mean much to them either. I don't think this adds any new information, only more numbers.
- For displaying codecs, maybe strip it down to just the decoder name instead of also showing the superfluous description? It's easily the thing using up the most space for no good reason
- Put putting audio sample rate and bitrate on the same row.

kasper93 commented:
>     * Consider renaming "Video Out" to just Display or something

Changed to `Display:`,

>     * Maybe only list specified and estimated display fps and file fps if they differ?

Added some logic to merge those if needed.

>     * I don't really think it's necessary to show bit depth information for pixel formats. You can already deduce this just from the pixel format itself, so if they can't then "bpp" isn't going to mean much to them either. I don't think this adds any new information, only more numbers.

I'm not sure about that, we can remove it.

>     * For displaying codecs, maybe strip it down to just the decoder name instead of also showing the superfluous description? It's easily the thing using up the most space for no good reason

Fixed it to not do nested desc with name. Now should be better.

>     * Put putting audio sample rate and bitrate on the same row.

I don't touch audio atm. Also I think for symmetry with video it should have it's own line.

@llyyr ^

> Perhaps "bpp" should only be showed when using hwdec. The formats of hwdec can sometimes be confusing, while the formats of swdec are always straightforward.

@arm64-v9a Not sure about that. For hwdec it already adds more to the lines, if anything it should be removed. I don't want some logic to should it only sometimes. Either we remove it or show always. I agree that for most formats it is obvious, so maybe indeed it is not needed.

kasper93 commented:
@llyyr @arm64-v9a @Hrxn Ok, I'm out of ideas for now. Take a look now and let me know what do you think.

llyyr commented:
There's some refactoring you can do but I believe they should be their own PR. This one has already gotten massive so it's fine to stop here I think before it gets too annoying to review for others.

Just some small nitpicks to things added in this PR: 

- We don't need to show [x:0, y:0, w:0, h:0] when no video-crop is inserted
- Frame and Bitrate can be on the same row to save some space.

Otherwise, this looks good to me visually at least. I'll go over the code in a bit

kasper93 commented:
> There's some refactoring you can do but I believe they should be their own PR.

Yes, I noticed few things while going through the code. But I need to stop or it will never be approved.

> We don't need to show [x:0, y:0, w:0, h:0] when no video-crop is inserted

This is blocked by https://github.com/mpv-player/mpv/pull/12307 I can pull the commit from there, but I hope at least this one will be merged soon.

> Frame and Bitrate can be on the same row to save some space.

Frame in it's full form can be `Frame: 3213 / 312412 (P) Interlaced", not sure I want add more there. For me bitrate doesn't fit anywhere if I'm being honest.

kasper93 commented:
I will remove bpp inferred from pixel format as this doesn't mean much, but for color-depth I would like to still keep it, because it is the target depth that libplacebo uses (and dither to), regardless of pixel format. See:
![image](https://github.com/mpv-player/mpv/assets/1126053/5ab9de47-7e7a-4116-830f-2285adac879b)

llyyr commented:
I preferred the old (8b) one than it having its own label, people shouldn't be watching stuff with stats.lua toggled on but we should still avoid clutter as much as possible

kasper93 commented:
Removed.

sfan5 commented on options/m_property.h:
```diff
@@ -107,6 +107,9 @@ struct m_property_action_arg {
 };
 
 enum mp_property_return {
+    // Returned from validator if action should be executed.
```
is this 'validator' concept an existing thing or did you just invent it?

kasper93 replied:
Yes, it is new concept.

kasper93 replied:
Just to expand, I wanted to avoid duplication of switch with the same cases for this type of property. So "validation" is a way to early exit when we don't actually going to return any data anyway.

Without this patch, single `mp.get_property_native`  produces 3 `VOCTRL` while 2 of them are completely unnecessary. And we does multiple properties we reduce the pressure by factor of 3.

sfan5 commented on player/lua/stats.lua:
```diff
@@ -721,114 +736,196 @@ local function pq_eotf(x)
-                         no_prefix_markup=true, suffix=")"}, {no=true, [""]=true})
+    local indent = o.prefix_sep .. o.prefix_sep
+
+    local pixel_format = r["hw-pixelformat"] or r["pixelformat"]
+    append(s, pixel_format, {prefix="Pixel Format:"})
```
didn't these used to get displayed alongside?

kasper93 replied:
Yes, it is explained in the commit msg. 

Basically mpv uses "fake" imgfmt for hwdec, which are used internal to track the format. But it is not real pixel format, it doesn't mean anything. For the fact that hwdec is enabled we display `hwdec-current`.

sfan5 commented on player/lua/stats.lua:
```diff
@@ -721,114 +736,196 @@ local function pq_eotf(x)
+    end
+    local has_prefix = false
+    if append_property(s, "estimated-frame-number", {prefix="Frame:"}) then
+        append_property(s, "estimated-frame-count", {indent=" / ", nl="",
```
mpv has no accurate frame numbers, this should never be displayed because users WILL rely on it

kasper93 replied:
I agree. It was request from @arm64-v9a, what do you think?

kasper93 replied:
For me even close frame number is good enough to know where to look at with other tools. Also it should be pretty accurate for sane video.

But I agree if the worry is that it is confusing for users, we can remove it.

Hrxn replied:
> mpv has no accurate frame numbers, this should never be displayed because users WILL rely on it

But which player has accurate frame numbes? 🤔

kasper93 replied:
@sfan5: How about I add option to stats.lua to show this estimated frame numbers? Disabled by default. If users enable it, it is their fault.

llyyr replied:
That sounds good to me

sfan5 replied:
Option + disabled by default is fine by me.

kasper93 replied:
> That sounds good to me, that option should also include all the other frame information such as frame type and interlacing

The main issue are estimated parameters, others aren't. But sure I can skip it, just displaying "B Interlaced" is not that useful. With frame number it makes whole package.

sfan5 commented on player/lua/stats.lua:
```diff
@@ -721,114 +736,196 @@ local function pq_eotf(x)
+                                  or {prefix="Picture Type:"}
+        append(s, frame_info["picture-type"], attrs)
+        has_prefix = true
+    end
```
I also argue that displaying this is not useful due to:
> Note that if any of these are used on OSD, the information might be off by a few frames due to OSD redrawing and frame display being somewhat disconnected, and you might have to pause and force a redraw.

It will also be out of date due to refresh interval of the stats.

kasper93 replied:
I think it is obvious that stats are not updated every frame. And it is perfectly fine to pause if you want to inspect single frame in more details. Same applies to peak detection value.

kasper93 replied:
I generally frame step, which works really well.

kasper93 commented on video/mp_image.h:
```diff
@@ -43,6 +43,7 @@
 // usually copy the whole struct, so that fields added later will be preserved.
 struct mp_image_params {
     enum mp_imgfmt imgfmt;      // pixel format
+    const char *imgfmt_name;    // pixel format name
```
> why put this here when it can be looked up from imgfmt?

Unfortunately it is not possible. We get `pl_fmt` from libplacebo, which is in no sane way mappable directly to imgfmt or even pixfmt. There are some fourcc's that could be used for mapping, but the list isn't exhaustive.

I mention that in first comment, I don't like the amount of glue this needs to talk to libplacebo. And frankly the best way forward would be to unify the structures between both projects to avoid constant and needles mapping of them.

This value is used to pass the name to the property, shouldn't be used in any other scenario.

>  also raises the question whether the caller guarantees that it points to permanently valid memory

This is on caller to ensure.

sfan5 replied:
so `imgfmt` is unset and `imgfmt_name` contains useful data, within the same structure?
that is confusing

kasper93 replied:
Alternative is duplicate everything, would be even more confusing.

sfan5 replied:
> This value is used to pass the name to the property, shouldn't be used in any other scenario.

this should be made clear then

kasper93 replied:
made it clear in commit msg

Dudemanguy commented on video/out/vo_gpu_next.c:
```diff
@@ -1461,18 +1487,25 @@ static int control(struct vo *vo, uint32_t request, void *data)
+        copy_hdr_metadata_to_mp(&p->last_hdr_metadata, data);
+        return true;
+
+    case VOCTRL_IMAGE_OUT_PARAMS: {
```
Do you need this? I confess I didn't look very closely but can't you just get the frame from the vo and then get the image from that?

kasper93 replied:
The vo_frame is actually the input to render, what I want to get is the target parameters of the render. It doesn't fit into vo_frame.

In properties terms
`video-target-params` -> target display params
`video-out-params` -> video out params. In fact it is "video in" (video out after filter chain), it is the input frame that VO holds and renders
`video-dec-params` -> video params after decode

I could hold something like target params in `struct vo`, but I think `VOCTRL` is more suited here.

As always, good ideas welcome. I don't know every inch of mpv :)

Dudemanguy replied:
Oh hmm, I can't say I have any better ideas. This is indeed information only the specific VO would know.

kasper93 commented:
@arm64-v9a: Unrelated to PR, but what font (and settings) are you using?

kasper93 commented:
Rebased only.

kasper93 commented:
Added `pl_frames_infer_mix`

haasn commented on video/out/vo_gpu_next.c:
```diff
@@ -1115,6 +1115,10 @@ static void draw_frame(struct vo *vo, struct vo_frame *frame)
     }
 
+#if PL_API_VER >= 330
+    pl_frames_infer_mix(p->rr, &mix, &target, NULL);
```
I wonder if we could also figure out a way to readback the inferred image parameters, which would finally solve those complains about DoVi metadata not being exposed via properties

kasper93 replied:
> DoVi metadata not being exposed via properties

What DV metadata you mean exactly? I can take a look, hmm...

kasper93 replied:
Actually nvm, you are right I can remove custom code for getting ref frame.

kasper93 replied:
done

kasper93 commented on player/command.c:
```diff
@@ -2278,12 +2278,12 @@ static int property_imgparams(struct mp_image_params p, int action, void *arg)
-    if (!!(desc.flags & MP_IMGFLAG_ALPHA) != (p.alpha != MP_ALPHA_AUTO)) {
-        p.alpha =
-            (desc.flags & MP_IMGFLAG_ALPHA) ? MP_ALPHA_STRAIGHT : MP_ALPHA_AUTO;
-    }
```
Which one exactly? 

```
int fmt = p.hw_subfmt ? p.hw_subfmt : p.imgfmt;
if (fmt) {
```

If we don't have `fmt` it will not be infered. `vo_gpu_next` doesn't set `fmt`, because there is no good way to map it, instead it sets `imgfmt_name` and `alpha` directly. It does not set `average-bpp`, I have a patch for that, but it is not that of a useful value anyway...

haasn commented:
Well, I am a bit unhappy introducing all this code duplication. I was thinking we could wait until after #12002 is done and libplacebo is made mandatory dependency, and instead add the pl colorspace fields everywhere

kasper93 commented:
> Well, I am a bit unhappy introducing all this code duplication.

What duplication exactly? There are two mapping functions added and that all.

> wait until after https://github.com/mpv-player/mpv/pull/12002 is done and libplacebo is made mandatory dependency, and instead add the pl colorspace fields everywhere

I would agree if it wasn't essentially mean to wait indefinitely. I don't expect this to happen soon. And frankly it is quite unrelated.

llyyr commented:
Perhaps the stats.lua parts of this can be merged, leaving the parts that require libplacebo code duplication to be done after we've made it mandatory?

kasper93 commented:
I can skip some changes if someone agrees to merge the others changes, but I will not start rebasing everything, just to make it rot for another few months, not worth my time.

Also without `VOCTRL_IMAGE_OUT_PARAMS` this split is not that much useful. There are lots of side-tracked cosmetic changes, but all depends on the new layout of the page in fact.

dyphire commented:
LGTM

kasper93 commented:
> Can you rebase to master?

Sure, pushed.

kasper93 commented:
let's make it a draft in favor of merging #12523 first

kasper93 commented:
> A-V Sync sometimes changes to scientific notation?

Yes, it is formatted with `%+.2g`, I like the properties of it. I can clamp all values `< 1e-5` to zero tho, with video-sync=audio it is just calc error with showing 3e-6. Though

kasper93 commented:
>  Problems remain. 1e-5 still seems insufficient.

This is valid value, it is just small.

kasper93 commented:
> Well, I am a bit unhappy introducing all this code duplication. I was thinking we could wait until after #12002 is done and libplacebo is made mandatory dependency, and instead add the pl colorspace fields everywhere

Does this patchset is really that bad to be blocked by not so near future API changes? As we can see libplacebo change is not coming anytime soon, so maybe we could unblock this PR? I can improve it, if there are particular pain point.

haasn commented:
> > Well, I am a bit unhappy introducing all this code duplication. I was thinking we could wait until after #12002 is done and libplacebo is made mandatory dependency, and instead add the pl colorspace fields everywhere
> 
> @haasn: Does this patchset is really that bad to be blocked by not so near future API changes? As we can see libplacebo change is not coming anytime soon, so maybe we could unblock this PR? I can improve it, if there are a particular pain points.

What is blocking #12002?

kasper93 commented:
> What is blocking #12002 / #12431?

I don't know. I don't merge things here. I think it looks ok, there might be meson wrap annoyance, but we can iterate on it once we identify the problems.

kasper93 commented:
@haasn: I cleared this PR as much as we can now. The rest is outside of this PR scope.

`pl_frames_infer_mix` change is probably blocked by https://code.videolan.org/videolan/libplacebo/-/issues/313 but the reset can be merged as-is. Thoughts?

kasper93 commented:
Now that `pl_hdr_metadata` is included in `mp_colorspace` and therefore in `video-out-params` should we remove `hdr-metadata` property? Nothing bad with it, just duplicated.

kasper93 commented:
Rebased on top of #12802 for simplicity. Only last 5 commits are related to this PR.

jeeb reviewed (approved):
LGTM.

4/5 moving the mix searching into the locking did confuse me for a while as I did not see a reason to move such "not just setting the parameters" action into the "hot path", but the last commit moves it outside. And I guess that commit is unlikely to get reverted so I will not go out of my way and edit the previous commit to move the search outside of the locked area :) .

And I guess this warning regarding `pl_frames_infer_mix` does not apply to gpu-next?
```c
// Warning: This does *not* call pl_frame.acquire/release, and so the returned
// metadata *may* be incorrect if the acquire callback mutates the pl_frame in
// nontrivial ways, in particular the crop and color space fields.
```

**edit**: Also something to ponder for the future, whether we want to see whether passing target parameters to VO reconfig makes sense, as with this change set VO parameters and VO target parameters are allocated in different places and at different points of time.
